service/matching/version_graph.go

```diff
@@ -0,0 +1,223 @@
+	}
+	return &workflowservice.GetWorkerBuildIdOrderingResponse{
+		CurrentDefault:   curDefault,
+		CompatibleLeaves: g.GetCompatibleLeaves(),
+	}
+}
+
+// Given an existing graph and an update request, update the graph appropriately.
+//
+// See the API docs for more detail. In short, the graph looks like one long line of default versions, each of which
+// is incompatible with the previous, optionally with branching compatibility branches. Like so:
+//
+// ─┬─1.0───2.0─┬─3.0───4.0
+//  │           ├─3.1
+//  │           └─3.2
+//  ├─1.1
+//  ├─1.2
+//  └─1.3
+//
+// In the above graph, 4.0 is the current default, and [1.3, 3.2] is the set of current compatible leaves. Links
+// going left are incompatible relationships, and links going up are compatible relationships.
+//
+// A request may:
+// 1. Add a new version to the graph, as a default version
+// 2. Add a new version to the graph, compatible with some existing version.
+// 3. Add a new version to the graph, compatible with some existing version and as the new default.
+// 4. Unset a version as a default. It will be dropped and its previous incompatible version becomes default.
+// 5. Unset a version as a compatible. It will be dropped and its previous compatible version will become the new
+//    compatible leaf for that branch.
+func UpdateVersionsGraph(
+	existingData *persistence.VersioningData,
+	req *workflowservice.UpdateWorkerBuildIdOrderingRequest,
+) error {
+	if req.GetVersionId().GetWorkerBuildId() == "" {
+		return serviceerror.NewInvalidArgument(
+			"request to update worker build id ordering is missing a valid version identifier")
+	}
+	// If the version is to become the new default, add it to the list of current defaults, possibly replacing
+	// the currently set one.
+	if req.GetBecomeDefault() {
+		curDefault := existingData.GetCurrentDefault()
+		if curDefault != nil {
+			// If the current default is going to be the previous compat version with the one we're adding,
+			// then we need to skip over it when setting the previous *incompatible* version.
+			if req.GetPreviousCompatible().GetWorkerBuildId() == curDefault.GetVersion().GetWorkerBuildId() {
+				existingData.CurrentDefault = &taskqueuepb.VersionIdNode{
+					Version:              req.VersionId,
+					PreviousCompatible:   curDefault,
+					PreviousIncompatible: curDefault.PreviousIncompatible,
+				}
+			} else {
+				// Otherwise, set the previous incompatible version to the current default.
+				existingData.CurrentDefault = &taskqueuepb.VersionIdNode{
+					Version:              req.VersionId,
+					PreviousCompatible:   nil,
+					PreviousIncompatible: curDefault,
+				}
+			}
+		} else {
+			if req.GetPreviousCompatible() != nil {
+				return serviceerror.NewInvalidArgument("adding a new default version which is compatible " +
+					" with some previous version, when there is no existing default version, is not allowed." +
+					" There must be a default version before this operation makes sense")
+			}
```

Comment:
Only if you're trying to add a version that you both want to be the default, and is supposed to be compatible with some existing version.

Since it wouldn't really make sense to have a node which isn't the default and isn't connected to anything else (the one being targeted as the previous compatible version with the current request)